### PR TITLE
[DEV-11525] Implement `Embed` ⇆ `WebBookmark` presentations transform

### PR DESCRIPTION
**Pull request by MohammadxAli**
https://github.com/prezly/slate/assets/67554982/8f758449-d1e8-4920-9c7e-dfe354d7c8db

### CODE REVIEW

File: packages/slate-editor/src/extensions/embed/components/EmbedMenu.tsx [hunk 1]
```diff
@@ -107,6 +120,16 @@ export function EmbedMenu({
                 </Toolbox.Section>
             )}
 
+            <Toolbox.Section caption="Change to...">
+                <OptionsGroup
+                    name="presentation"
+                    options={PRESENTATION_OPTIONS}
+                    selectedValue={EmbedNode.Presentation.EMBED}
+                    onChange={(presentation) => onTransform(presentation)}
```
Comment (e1himself) [inline packages/slate-editor/src/extensions/embed/components/EmbedMenu.tsx:128]: Is there something preventing from doing this?

```suggestion
                    onChange={onTransform}
```

File: packages/slate-editor/src/extensions/embed/components/EmbedMenu.tsx [hunk 2]
```diff
@@ -59,11 +60,23 @@ const LAYOUT_OPTIONS: OptionsGroupOption<EmbedNode.Layout>[] = [
     },
 ];
 
+const PRESENTATION_OPTIONS: OptionsGroupOption<EmbedNode.Presentation>[] = [
+    {
+        value: EmbedNode.Presentation.EMBED,
```
Comment (e1himself) [inline packages/slate-editor/src/extensions/embed/components/EmbedMenu.tsx:65]: I suggest using `EmbedNode.TYPE` and `BookmarkNode.TYPE` here. It's more straightforward, and you won't need a new custom enum (located in a surprizing place) :)

File: packages/slate-editor/src/extensions/web-bookmark/components/WebBookmarkMenu.tsx [hunk 3]
```diff
@@ -42,10 +43,22 @@ const LAYOUT_OPTIONS: OptionsGroupOption<BookmarkCardLayout>[] = [
...
+const PRESENTATION_OPTIONS: OptionsGroupOption<BookmarkNode.Presentation>[] = [
+    {
+        value: BookmarkNode.Presentation.BOOKMARK,
+        label: 'Bookmark',
+    },
+    {
+        value: BookmarkNode.Presentation.EMBED,
+        label: 'Embed',
+    },
+];
```
Comment (e1himself) [inline packages/slate-editor/src/extensions/web-bookmark/components/WebBookmarkMenu.tsx:55]: The order of these options should always match in all menus. It has to be: `Embed`, `Bookmark`, `Link` (when it's available)

File: packages/slate-editor/src/extensions/embed/components/EmbedElement.tsx [hunk 4]
```diff
@@ -48,6 +48,12 @@ export function EmbedElement({
     const handleRemove = useCallback(() => {
         removeEmbed(editor, element);
     }, [editor, element]);
+    const handleTransform = useCallback(
+        (presentation: `${EmbedNode.Presentation}`) => {
+            transformEmbed(editor, element, presentation);
+        },
+        [editor, element],
+    );
```
[the remark this answers is not in the document]
Comment (MohammadxAli) [reply packages/slate-editor/src/extensions/embed/components/EmbedElement.tsx:56]: Agree! Great suggestion 👍

File: packages/slate-editor/src/extensions/embed/components/EmbedMenu.tsx (hunk 2 above)
Comment (MohammadxAli) [reply packages/slate-editor/src/extensions/embed/components/EmbedMenu.tsx]: Something weird happens with mocha when I'm trying to use `EmbedNode.TYPE` in `WebBookmarkMenu`:

```log
@prezly/slate-editor:test: > @prezly/slate-editor@0.94.2 test:unit /home/runner/work/slate/slate/packages/slate-editor
@prezly/slate-editor:test: > mocha --exit --config ../../.mocha/config.cjs './src/**/*.test.{ts,tsx}'
@prezly/slate-editor:test: 
@prezly/slate-editor:test: 
@prezly/slate-editor:test: TypeError: Cannot read properties of undefined (reading 'EmbedNode')
@prezly/slate-editor:test:     at Object.EmbedNode (/home/runner/work/slate/slate/packages/slate-editor/src/extensions/embed/index.ts:21:23)
@prezly/slate-editor:test:     at Object.<anonymous> (/home/runner/work/slate/slate/packages/slate-editor/src/extensions/web-bookmark/components/WebBookmarkMenu.tsx:51:16)
```

I removed the import of `EmbedNode` here:

https://github.com/prezly/slate/pull/490/commits/003d705847d567fc1420a870357bb773ff242fbb

And used `"embed"` directly and the type check command works.

---

Here's a failed build:

https://github.com/prezly/slate/actions/runs/6038947737/job/16386466296

Comment (MohammadxAli) [reply packages/slate-editor/src/extensions/embed/components/EmbedMenu.tsx]: Also, I have converted the `BookmarkNode` types to use `namespace`:

https://github.com/prezly/slate/pull/490/commits/8ca298fe8ed6facd6b3c5e6b7075f4a9f9f75007

Comment (e1himself) [reply packages/slate-editor/src/extensions/embed/components/EmbedMenu.tsx]: > Also, I have converted the BookmarkNode types to use namespace:

Great! :+1: 

> Something weird happens with mocha when I'm trying to use EmbedNode.TYPE in WebBookmarkMenu:

Circular dependency it is!

So using the node types here is not an option, it seems. :thinking: 

Let's use a local type then? Something like: 

```ts
export type Presentation = 'embed' | 'card';
```

File: packages/slate-editor/src/extensions/web-bookmark/types.ts [hunk 5]
```diff
@@ -1,6 +1,12 @@
...
+
+import type { EmbedNode } from '#extensions/embed';
 
 export interface WebBookmarkExtensionParameters {
     withNewTabOption?: boolean;
+    withConversionOptions?: boolean;
     fetchOembed: (url: OEmbedInfo['url']) => Promise<OEmbedInfo>;
 }
+
+export type Presentation = `${typeof EmbedNode.TYPE}` | `${typeof BookmarkNode.TYPE}`;
```
Comment (e1himself) [inline packages/slate-editor/src/extensions/web-bookmark/types.ts:12]: If using the type values everywhere is not an option, let's disconnect them then:

```suggestion
export type Presentation = 'embed' | 'card';
```

File: packages/slate-editor/src/extensions/embed/types.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import type { BookmarkNode } from '@prezly/slate-types';
+
+import type { EmbedNode } from './EmbedNode';
+
+export type Presentation = `${typeof EmbedNode.TYPE}` | `${typeof BookmarkNode.TYPE}`;
```
Comment (e1himself) [inline packages/slate-editor/src/extensions/embed/types.ts:5]:
```suggestion
export type Presentation = `'embed' | 'card';
```

**Review by e1himself — APPROVED** [review]
Looks great! 

Please see a few minor remarks above, and good to go :+1:

**Pull request merged**